lnrpc/routerrpc/router_server.go

```diff
@@ -404,8 +449,241 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+
+		// Dispatch the payment probe with adjusted fee amount.
+		resp, err := s.sendProbePayment(ctx, probeRequest)
+		if err != nil {
+			return nil, err
+		}
+		resp.Dest = routeHintDest[:]
+		resp.ChannelId = hopHint.ChannelID
+		resp.TimeLockDelay = int64(hopHint.CLTVExpiryDelta)
+
+		// The probe succeeded, so we can add the last hop's fee to the
+		// to fee the payment probe returned.
+		resp.RoutingFeeMsat += hopFee
+
+		// Record the fee estimate in our overall response list.
+		feeEstimates = append(feeEstimates, resp)
+	}
+
+	return feeEstimates, nil
+}
+
+// probePaymentStream is a custom implementation of the grpc.ServerStream
+// interface. It is used to send payment status updates to the caller on the
+// stream channel.
+type probePaymentStream struct {
+	Router_SendPaymentV2Server
+
+	stream chan *lnrpc.Payment
+	ctx    context.Context //nolint:containedctx
+}
+
+// Send sends a payment status update to a payment stream that the caller can
+// evaluate.
+func (p *probePaymentStream) Send(response *lnrpc.Payment) error {
+	p.stream <- response
+
+	return nil
+}
+
+// Context returns the context of the stream.
+func (p *probePaymentStream) Context() context.Context {
+	return p.ctx
+}
+
+// sendProbePayment sends a payment to a target node in order to obtain
+// potential routing fees for it. The payment request has to contain a payment
+// hash that is guaranteed to be unknown to the target node, so it cannot settle
+// the payment. This method invokes a payment request loop in a goroutine and
+// awaits payment status updates.
+func (s *Server) sendProbePayment(ctx context.Context,
+	req *SendPaymentRequest) (*DestinationFeeEstimate, error) {
+
+	// We'll launch a goroutine to send the payment probes.
+	errChan := make(chan error, 1)
+	defer close(errChan)
+
+	paymentStream := &probePaymentStream{
+		stream: make(chan *lnrpc.Payment),
+		ctx:    ctx,
+	}
+	go func() {
+		err := s.SendPaymentV2(req, paymentStream)
+		if err != nil {
+			errChan <- err
+		}
+	}()
+
+	// paymentFee returns the fee of the last attempt of the payment.
+	paymentFee := func(p *lnrpc.Payment) int64 {
+		if len(p.Htlcs) > 0 {
+			lastAttempt := p.Htlcs[len(p.Htlcs)-1]
+			if lastAttempt.Route != nil {
+				return lastAttempt.Route.TotalFeesMsat
+			}
+		}
+
+		return 0
+	}
+
+	for {
+		select {
+		case payment := <-paymentStream.stream:
+			switch payment.Status {
+			case lnrpc.Payment_INITIATED:
+			case lnrpc.Payment_IN_FLIGHT:
+				continue
+
+			case lnrpc.Payment_SUCCEEDED:
+				return nil, errors.New("probe payment " +
+					"unexpectedly succeeded")
+
+			case lnrpc.Payment_FAILED:
+				// Incorrect payment details points to a
+				// successful probe.
+				if payment.FailureReason == lnrpc.PaymentFailureReason_FAILURE_REASON_INCORRECT_PAYMENT_DETAILS { //nolint:lll
+					return &DestinationFeeEstimate{
+						RoutingFeeMsat: paymentFee(
+							payment,
+						),
+						FailureReason: payment.FailureReason, //nolint:lll
```

Comment:
Should this be `FAILURE_REASON_NONE`? And the other fields in the DestinationFeeEstimate should be set? `time_lock_delay`, `dest`, and `channel_id`.